File: LeetCode/50_Pow(x, n).cpp
class Solution {
public:
    double myPow(double x, int n) {
        if (n == 0) {
            return 1;
        }

        long remain = n;
        bool reverse = false;
        if (remain < 0) {
            remain = -remain;
            reverse = true;
        }

        double ans = 1;
        double temp;
        long times;

        while (remain > 1) {
            temp = x * x;
            times = 2;
            do {
                long test = times << 1;
                if (test > remain) {
                    break;
                }
                temp *= temp;
                times = test;
            } while (true);

            ans *= temp;
            remain -= times;
        }
        if (remain == 1) {
            ans *= x;
        }

        return (reverse == false)? ans : 1 / ans;
    }
};

File: LeetCode/50_Pow(x, n).java
class Solution {
    public double myPow(double x, int n) {

        long exp = n;
        if (exp < 0) {
            exp = -exp;
            return 1 / fastPow(x, exp);
        }
        return fastPow(x, exp);
    }

    private double fastPow(double x, long exp) {

        if (exp == 0) {
            return 1;
        }

        double half = fastPow(x, exp >> 1);
        double result = half * half;

        if (exp % 2 == 1) {
            result *= x;
        }

        return result;
    }
}